### PR TITLE
Feat/itunes-tunecard UI

### PR DESCRIPTION
**Pull request by DebarshiB13**
### Ticket Link

---

### Related Links

---

### Description

---
Second PR for the Itunes app.

The new changes are :
Implementation of the UI part.
List of songs displayed from search results in form of Grid
Each song has a card.
Play/Stop songs
Unit tests for each item.

### Steps to Reproduce / Test

---
 * Go to localhost:3000/songs.
 * Type any song or artist name on the search bar.
 * List of cards/ error will appear based on the response.
 * Play/stop any song by clicking the play/stop button.

### GIF's

---
feat_tunecard_ui
feat_tunecard_ui_test

### CODE REVIEW

File: app/components/TuneCard/tests/index.test.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  it('should play/pause audio when button is clicked', () => {
+    const playStub = jest.spyOn(window.HTMLMediaElement.prototype, 'play').mockImplementation(() => {});
+    const pauseStub = jest.spyOn(window.HTMLMediaElement.prototype, 'pause').mockImplementation(() => {});
+
+    const { getByTestId } = renderWithIntl(<TuneCard />);
+
+    const button = getByTestId('play-pause-btn');
+
+    fireEvent.click(button);
```
Comment (alichherawalla) [inline app/components/TuneCard/tests/index.test.js]: just so its amply clear can we do this 
```
expect(playStub).not.toHaveBeenCalled();
fireEvent.click(button);
expect(playStub).toHaveBeenCalled();

expect(playStub).not.toHaveBeenCalled();
fireEvent.click(button);
expect(pauseStub).toHaveBeenCalled();
```

File: app/containers/TuneContainer/index.js [hunk 2]
```diff
@@ -69,14 +89,76 @@ export function TuneContainer({
...
+  const renderSongList = () => {
+    const results = get(songsData, 'results', []);
+    const resultCount = get(songsData, 'resultCount', 0);
+
+    return (
+      <If condition={results.length !== 0 || loading}>
+        <CustomCard>
+          <Skeleton loading={loading} active>
+            <If condition={searchTerm}>
+              <div>
```
Comment (alichherawalla) [inline app/containers/TuneContainer/index.js]: please remove extra nesting

File: app/components/TuneCard/tests/index.test.js (hunk 1 above)
Comment (DebarshiB13) [reply app/components/TuneCard/tests/index.test.js]: Removed the test as the implementation has changed in next commit

File: app/components/TuneCard/index.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      <Image src={cardImg} width="100%" preview={false} />
+      <Typography.Paragraph style={{ fontSize: 18 }}>{collectionName}</Typography.Paragraph>
+
+      <IconButton
+        shape="circle"
+        type="text"
+        data-testid="play-pause-btn"
+        onClick={() => handlePlayPause(previewUrl)}
+        icon={
+          currentTrack?.src === previewUrl ? (
```
Comment (alichherawalla) [inline app/components/TuneCard/index.js]: use `If` it will read slightly better

File: app/components/TuneCard/index.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+export function TuneCard({
+  maxwidth,
+  artistName,
+  collectionName,
+  cardImg,
+  previewUrl,
+  currentTrack,
+  setCurrentTrack,
+  isTrackPlaying,
```
Comment (alichherawalla) [inline app/components/TuneCard/index.js]: why don't we delegate the playing/pausing of the track to the component itself. Looks like we're controlling it based on some prop values that are passed from outside.

File: app/components/TuneCard/index.js (hunk 3 above)
Comment (DebarshiB13) [reply app/components/TuneCard/index.js]: Fixed in next commit

File: app/components/TuneCard/index.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  display: flex;
+  justify-content: center;
+  align-items: center;
+`;
+
+export function TuneCard({ maxwidth, artistName, collectionName, cardImg, previewUrl, handleOnActionClick }) {
+  const audioRef = useRef();
+  const [isPlaying, setIsPlaying] = useState(false);
+
+  useEffect(() => {
```
Comment (alichherawalla) [inline app/components/TuneCard/index.js]: this is very inefficient, even if its false your constantly setting it as false. Do we really need to do this? 

we can completely remove the need of the `isPlaying` state right?

File: app/components/TuneCard/tests/index.test.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ */
+
+import React from 'react';
+import { fireEvent } from '@testing-library/dom';
+import { renderWithIntl, timeout } from '@utils/testUtils';
+import TuneCard from '../index';
+
+describe('<TuneCard />', () => {
+  it('should render and match the snapshot', () => {
```
Comment (alichherawalla) [inline app/components/TuneCard/tests/index.test.js:28]: add a test to make sure that everything is artistName, collectioName is getting rendered correctly

File: app/containers/TuneContainer/index.js [hunk 7]
```diff
@@ -70,15 +90,87 @@ export function TuneContainer({
   };
 
   const debouncedHandleOnChange = debounce(handleOnChange, 200);
+
+  const renderSongList = () => {
+    const results = get(songsData, 'results', []);
+    const resultCount = get(songsData, 'resultCount', 0);
+
+    const handleOnActionClick = (ref, isPlaying) => {
+      // console.log({ ref });
```
Comment (alichherawalla) [inline app/containers/TuneContainer/index.js]: remove console

File: app/containers/TuneContainer/index.js [hunk 8]
```diff
@@ -70,15 +90,87 @@ export function TuneContainer({
...
+      }
+    };
+    return (
+      <If condition={results.length !== 0 || loading} otherwise={null}>
+        <CustomCard>
+          <Skeleton loading={loading} active>
+            <If condition={searchTerm} otherwise={null}>
+              <T id="search_term" values={{ searchTerm }} />
+            </If>
+            <If condition={resultCount !== 0}>
```
Comment (alichherawalla) [inline app/containers/TuneContainer/index.js]:
```
  <If condition={resultCount}>
```
resultCount is truthy

File: app/containers/TuneContainer/index.js [hunk 9]
```diff
@@ -70,15 +90,87 @@ export function TuneContainer({
...
+              of={results}
+              renderItem={(result, index) => (
+                <TuneCard
+                  key={index}
+                  handleOnActionClick={handleOnActionClick}
+                  artistName={result.artistName}
+                  collectionName={result.collectionName}
+                  cardImg={result.artworkUrl100}
+                  previewUrl={result.previewUrl}
+                  currentTrack={currentTrack}
```
Comment (alichherawalla) [inline app/containers/TuneContainer/index.js]: why do we need to pass currentTrack?

File: app/components/TuneCard/index.js (hunk 5 above)
Comment (DebarshiB13) [reply app/components/TuneCard/index.js]: Remove the state from TuneCard

File: app/containers/TuneContainer/index.js (hunk 7 above)
Comment (DebarshiB13) [reply app/containers/TuneContainer/index.js]: removed

File: app/containers/TuneContainer/index.js (hunk 8 above)
Comment (DebarshiB13) [reply app/containers/TuneContainer/index.js]: Updated accordingly

File: app/containers/TuneContainer/index.js (hunk 9 above)
Comment (DebarshiB13) [reply app/containers/TuneContainer/index.js]: removed

File: app/components/TuneCard/tests/index.test.js (hunk 6 above)
Comment (DebarshiB13) [reply app/components/TuneCard/tests/index.test.js:28]: Added in next commit

File: 0001-tests.patch [hunk 10]
```diff
@@ -0,0 +1,140 @@
+From 1217055b0dfd9eda3418496633d92fce76b4dc00 Mon Sep 17 00:00:00 2001
```
Comment (alichherawalla) [inline 0001-tests.patch]: please delete the patch file. Add a *.patch rule in the .gitignore as well

Comment (DebarshiB13) [reply 0001-tests.patch]: Updated accordingly.

File: app/components/TuneCard/index.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    handleOnActionClick(audioRef);
+  };
+
+  return (
+    <ItemCard maxwidth={maxwidth} data-testid="tune-card">
+      <Typography.Title style={{ fontSize: 16 }} data-testid="artist-name">
+        {artistName}
+      </Typography.Title>
+      <Image src={cardImg} width="100%" preview={false} />
+      <Typography.Paragraph style={{ fontSize: 18 }}>{collectionName}</Typography.Paragraph>
```
Comment (vishnu-wednesday) [inline app/components/TuneCard/index.js]: A very minor nit. Maybe we can move all these to styled-components and keep that consistency or even use the https://styled-components.com/docs/api#css-prop API

File: app/containers/TuneContainer/index.js [hunk 12]
```diff
@@ -16,18 +16,37 @@ import tuneContainerSaga from './saga';
 import { injectSaga } from 'redux-injectors';
 import get from 'lodash/get';
 import isEmpty from 'lodash/isEmpty';
-import { Input } from 'antd';
+import { Input, Card, Skeleton, Row } from 'antd';
 import styled from 'styled-components';
 import debounce from 'lodash/debounce';
+import T from '@components/T';
+import { TuneCard } from '@components/TuneCard/index';
+import For from '@components/For/index';
```
Comment (vishnu-wednesday) [inline app/containers/TuneContainer/index.js]: `@components/For`

File: app/containers/TuneContainer/index.js [hunk 13]
```diff
@@ -16,18 +16,37 @@ import tuneContainerSaga from './saga';
...
+import If from '@app/components/If/index';
```
Comment (vishnu-wednesday) [inline app/containers/TuneContainer/index.js]: Same as above

File: app/containers/TuneContainer/tests/index.test.js [hunk 14]
```diff
@@ -25,25 +26,45 @@ describe('<TuneContainer /> container tests', () => {
     expect(getByTestId('search-bar')).toBeInTheDocument();
   });
 
-  it('should set "Loading" to false and render "Loaded" when songsData or SongsError is available', async () => {
-    const songsData = {
-      resultCount: 1,
-      results: [{ id: 1, song: 'Some data' }]
+  it('should render for component when songsData is available', () => {
```
Comment (vishnu-wednesday) [inline app/containers/TuneContainer/tests/index.test.js]: should render `For` component

File: app/containers/TuneContainer/index.js [hunk 15]
```diff
@@ -70,15 +89,82 @@ export function TuneContainer({
...
+    const handleOnActionClick = (ref) => {
+      setCurrentTrack(ref);
+      const isPaused = currentTrack?.current?.paused;
+      if (!isPaused && ref?.current.src !== currentTrack?.current?.src) {
+        currentTrack?.current.pause();
+      }
+    };
+
+    return (
+      <If condition={results.length !== 0 || loading} otherwise={null}>
```
Comment (vishnu-wednesday) [inline app/containers/TuneContainer/index.js]: Do you need the `!==` check?

File: app/containers/TuneContainer/index.js (hunk 12 above)
Comment (DebarshiB13) [reply app/containers/TuneContainer/index.js]: Updated accordingly

File: app/containers/TuneContainer/index.js (hunk 13 above)
Comment (DebarshiB13) [reply app/containers/TuneContainer/index.js]: Updated accordingly

File: app/containers/TuneContainer/tests/index.test.js (hunk 14 above)
Comment (DebarshiB13) [reply app/containers/TuneContainer/tests/index.test.js]: Updated accordingly

File: app/containers/TuneContainer/index.js (hunk 15 above)
Comment (DebarshiB13) [reply app/containers/TuneContainer/index.js]: Updated accordingly

**Review by vishnu-wednesday — APPROVED** [review]

**Pull request merged**